### PR TITLE
Pwawrzen/tra 14 creating an api for communication

### PR DESCRIPTION
**Pull request by AnimaVillis**

### CODE REVIEW

File: backend/AuthMiddleware.php [hunk 1]
```diff
@@ -0,0 +1,63 @@
...
+                    "message" => $data['message'],
+                ];
+            endif;
+        } else {
+            return [
+                "success" => 0,
+                "message" => "Token not found in request"
+            ];
+        }
+    }
```
Comment (aleksanderwalczuk) [inline backend/AuthMiddleware.php:44]: This function returns void if conditions are unmet. It should return 0.

File: backend/api/add_mentee.php [hunk 2]
```diff
@@ -0,0 +1,38 @@
+<?php
+    header("Access-Control-Allow-Origin: *");
+    header("Content-Type: application/json; charset=UTF-8");
+    header("Access-Control-Allow-Methods: POST");
+    header("Access-Control-Max-Age: 3600");
+    header("Access-Control-Allow-Headers: Content-Type, Access-Control-Allow-Headers, Authorization, X-Requested-With");
+    //ini_set('display_errors', 1);
```
Comment (aleksanderwalczuk) [inline backend/api/add_mentee.php:7]: please do not leave commented code.

File: backend/api/add_mentee.php [hunk 3]
```diff
@@ -0,0 +1,38 @@
...
+    $item = new Mentee($db);
+    $data = json_decode(file_get_contents("php://input"));
+    if($data != null) {
+        $item->name = $data->name;
+        $item->email = $data->email;
+        $item->password = $data->password;
+        $item->user_level = $data->user_level;
+        $item->first_login = $data->first_login;
+        $item->account_activation = $data->account_activation;
+        if($item->createMentee()){
```
Comment (aleksanderwalczuk) [inline backend/api/add_mentee.php:21]: This is a user, not an item.

File: backend/api/add_mentee.php [hunk 4]
```diff
@@ -0,0 +1,38 @@
...
+            http_response_code(200);
+            echo json_encode(
+                array("message" => "Mentee added successfully.")
+            );
+        } else{
+            http_response_code(200);
+            echo json_encode(
+                array("message" => "Mentee cannot be added.")
+            );
```
Comment (aleksanderwalczuk) [inline backend/api/add_mentee.php:30]: Why response code is `200` 😮

File: backend/api/add_mentee.php [hunk 5]
```diff
@@ -0,0 +1,38 @@
...
+        }
+    } else {
+        http_response_code(200);
+        echo json_encode(
+            array("message" => "Fill all fields.")
+        );
```
Comment (aleksanderwalczuk) [inline backend/api/add_mentee.php:36]: Do you know what response `200` means?

File: backend/api/delete_mentee.php [hunk 6]
```diff
@@ -0,0 +1,31 @@
...
+    if($item->deleteMentee()){
+        http_response_code(200);
+            echo json_encode(
+                array("message" => "Mentee deleted.")
+            );
+    } else{
+        http_response_code(200);
+            echo json_encode(
+                array("message" => "Mentee doesn't exsist.")
+            );
```
Comment (aleksanderwalczuk) [inline backend/api/delete_mentee.php:29]: Please adjust response code.

File: backend/api/delete_mentee.php [hunk 7]
```diff
@@ -0,0 +1,31 @@
...
+    $db = $database->getConnection();
+    
+    $item = new Mentee($db);
+    
+    $data = json_decode(file_get_contents("php://input"));
+    
+    $item->id = $data->id;
+    
+    if($item->deleteMentee()){
+        http_response_code(200);
```
Comment (aleksanderwalczuk) [inline backend/api/delete_mentee.php:21]: This should be `204`

File: backend/api/mentee_first_login.php [hunk 8]
```diff
@@ -0,0 +1,40 @@
...
+            echo json_encode(
+                array("message" => "Mentee first login not send.")
+            );
+        }
+    } else {
+        http_response_code(200);
+        echo json_encode(
+            array("message" => "Fill all fields.")
+        );
+    }
```
Comment (aleksanderwalczuk) [inline backend/api/mentee_first_login.php:39]: Response codes again

File: backend/api/mentee_first_login.php [hunk 9]
```diff
@@ -0,0 +1,40 @@
...
+    $data = json_decode(file_get_contents("php://input"));
+    if($data != null) {
+        $item->users_id = $data->users_id;
+        $item->initial_weight = $data->initial_weight;
+        $item->current_weight = $data->current_weight;
+        $item->target_weight = $data->target_weight;
+        $item->growth = $data->growth;
+        $item->age = $data->age;
+        $item->activity_factor = $data->activity_factor;
+        $item->sex = $data->sex;
```
Comment (aleksanderwalczuk) [inline backend/api/mentee_first_login.php:22]: User

File: backend/api/one_mentee.php [hunk 10]
```diff
@@ -0,0 +1,34 @@
...
+    header("Access-Control-Allow-Headers: Content-Type, Access-Control-Allow-Headers, Authorization, X-Requested-With");
+
+    include_once '../config/database.php';
+    include_once '../class/mentee.php';
+    $database = new Database();
+    $db = $database->getConnection();
+    $item = new Mentee($db);
+    $item->id = isset($_GET['id']) ? $_GET['id'] : die();
+  
+    $item->getSingleMentee();
```
Comment (aleksanderwalczuk) [inline backend/api/one_mentee.php:15]: user

File: backend/api/one_mentee.php [hunk 11]
```diff
@@ -0,0 +1,34 @@
...
+    if($item->name != null){
+        // create array
+        $ment_arr = array(
```
Comment (aleksanderwalczuk) [inline backend/api/one_mentee.php:18]: `$users`

File: backend/api/recovery_mentee.php [hunk 12]
```diff
@@ -0,0 +1,35 @@
...
+    include_once '../class/mentee.php';
+    $database = new Database();
+    $db = $database->getConnection();
+    $item = new Mentee($db);
+    $data = json_decode(file_get_contents("php://input"));
+    if($data != null) {
+        $item->name = $data->name;
+        $item->email = $data->email;
+        if($item->recoveryMentee()){
+            //dodanie wysyłki maila
```
Comment (aleksanderwalczuk) [inline backend/api/recovery_mentee.php:18]: avoid mixing languages

File: backend/api/recovery_mentee.php [hunk 13]
```diff
@@ -0,0 +1,35 @@
...
+                array("message" => "Mentee recovery mail sended.")
+            );
+        } else{
+            http_response_code(200);
+            echo json_encode(
+                array("message" => "Mentee cannot be recovered.")
+            );
+        }
+    } else {
+        http_response_code(200);
```
Comment (aleksanderwalczuk) [inline backend/api/recovery_mentee.php:30]: Response codes....

File: backend/class/JwtHandler.php [hunk 14]
```diff
@@ -0,0 +1,62 @@
...
+    protected $jwt_secrect;
+    protected $token;
+    protected $issuedAt;
+    protected $expire;
+    protected $jwt;
+
+    public function __construct()
+    {
+        // set your default time-zone
+        date_default_timezone_set('America/Los_Angeles');
```
Comment (aleksanderwalczuk) [inline backend/class/JwtHandler.php:18]: Is this our timezone?

File: backend/class/JwtHandler.php [hunk 15]
```diff
@@ -0,0 +1,62 @@
...
+        $this->issuedAt = time();
+
+        // Token Validity (3600 second = 1hr)
+        $this->expire = $this->issuedAt + 3600;
+
+        // Set your secret or signature
+        $this->jwt_secrect = "MySecretIsMySecret";
```
Comment (aleksanderwalczuk) [inline backend/class/JwtHandler.php:25]: What's that?

File: backend/class/JwtHandler.php [hunk 16]
```diff
@@ -0,0 +1,62 @@
...
+        );
+
+        $this->jwt = JWT::encode($this->token, $this->jwt_secrect, 'HS256');
+        return $this->jwt;
+    }
+
+    public function jwtDecodeData($jwt_token)
+    {
+        try {
+            //$decode = JWT::decode($jwt_token, $this->jwt_secrect, array('HS256'));
```
Comment (aleksanderwalczuk) [inline backend/class/JwtHandler.php:50]: Commented code again 😮‍💨

File: backend/class/mentee.php [hunk 17]
```diff
@@ -0,0 +1,151 @@
...
+            $stmt->bindParam(":first_login", $this->first_login);
+            $stmt->bindParam(":account_activation", $this->account_activation);
+        
+            if($stmt->execute()){
+               return true;
+            }
+            return false;
+        }
+
+        public function firstloginMentee(){
```
Comment (aleksanderwalczuk) [inline backend/class/mentee.php:61]: Actually why is it called first login? 🤔

File: backend/class/mentee.php [hunk 18]
```diff
@@ -0,0 +1,151 @@
...
+            $stmt->bindParam(":sex", $this->sex);
+        
+            if($stmt->execute()){
+               return true;
+            }
+            return false;
+        }
+
+        public function getSingleMentee(){
+            $sqlQuery = "SELECT id, name, email, user_level, first_login FROM ". $this->db_table ." WHERE id = ? LIMIT 1";
```
Comment (aleksanderwalczuk) [inline backend/class/mentee.php:103]: Do we need `LIMIT` if we have an `id`?

File: backend/class/mentee.php [hunk 19]
```diff
@@ -0,0 +1,151 @@
...
+            
+            $this->id = $dataRow['id'];
+            $this->name = $dataRow['name'];
+            $this->email = $dataRow['email'];
+            $this->user_level = $dataRow['user_level'];
+            $this->first_login = $dataRow['first_login'];
+        }        
+
+        public function recoveryMentee(){
+            $sqlQuery = "SELECT name, email FROM ". $this->db_table ." WHERE name = :name AND email = :email LIMIT 1";
```
Comment (aleksanderwalczuk) [inline backend/class/mentee.php:117]: Do we need `LIMIT` if we have an `id`?

File: backend/class/mentee.php [hunk 20]
```diff
@@ -0,0 +1,151 @@
...
+            $this->name=htmlspecialchars(strip_tags($this->name));
+            $this->email=htmlspecialchars(strip_tags($this->email));
+            
+
+            if($dataRow['name'] == $this->name AND $dataRow['email'] == $this->email) {
+                if($stmt->execute()){
+                    return true;
+                } else {
+                    return false;
+                }
```
Comment (aleksanderwalczuk) [inline backend/class/mentee.php:133]:
```suggestion
                return false;
```

File: backend/config/database.php [hunk 21]
```diff
@@ -0,0 +1,19 @@
+<?php 
+    class Database {
+        private $host = "89.117.56.9";
+        private $database_name = "projekt";
+        private $username = "projektbaza";
+        private $password = "ViJbqNT0cFFHAkSoN8tm";
```
Comment (aleksanderwalczuk) [inline backend/config/database.php:6]: Those fields are insecure. We should store them as env variables.

File: backend/login.php [hunk 22]
```diff
@@ -0,0 +1,100 @@
...
+                    
+                    $returnData = [
+                        'success' => 1,
+                        'message' => 'You have successfully logged in.',
+                        'token' => $token
+                    ];
+
+                // IF INVALID PASSWORD
+                else:
+                    $returnData = msg(0,422,'Invalid Password!');
```
Comment (aleksanderwalczuk) [inline backend/login.php:84]: It's a bad request, `400`

File: backend/login.php [hunk 23]
```diff
@@ -0,0 +1,100 @@
...
+                endif;
+
+            // IF THE USER IS NOT FOUNDED BY EMAIL THEN SHOW THE FOLLOWING ERROR
+            else:
+                $returnData = msg(0,422,'Invalid Email Address!');
```
Comment (aleksanderwalczuk) [inline backend/login.php:89]: This is a classic `404`

File: frontend/pages/Login/Login.vue [hunk 24]
```diff
@@ -4,7 +4,7 @@
             <h1 class="text-3xl font-bold tracking-widest">SIGN UP</h1>
             <span>Don't have an account yet? Sign up here!</span>
         </div>
-            <form action="../backend/includes/signup.inc.php" method="post" class="flex flex-col items-center justify-center space-y-6">
+            <form class="flex flex-col items-center justify-center space-y-6">
```
Comment (aleksanderwalczuk) [inline frontend/pages/Login/Login.vue:7]: Should not be changed.

**Review by aleksanderwalczuk — CHANGES_REQUESTED** [review]
Vendor files (dependencies) should not be commited. Please remove them from this PR and respond in threads I've opened.

File: backend/class/mentee.php (hunk 17 above)
Comment (AnimaVillis) [reply backend/class/mentee.php:61]: The first login function is responsible for invoking the form needed to complete the information about the variables affecting the calculation of caloric demand.

Comment (aleksanderwalczuk) [reply backend/class/mentee.php:61]: Then it's onboarding, not first login 😉

**Review by aleksanderwalczuk — APPROVED** [review]

**Pull request merged**